File: upload_images.py
#!/usr/bin/env python3
"""
Génère des images PLACEHOLDER (fond + texte), pas de vraies photos produit.
Pour des photos réelles, utilisez: python sync_real_product_images.py
"""
import io
import os
import sys
from urllib.parse import quote

import requests
from PIL import Image, ImageDraw

SUPABASE_URL = os.environ.get("SUPABASE_URL", "").rstrip("/")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY") or os.environ.get("SUPABASE_KEY")
BUCKET_NAME = os.environ.get("SUPABASE_BUCKET", "products")

# Products data
PRODUCTS = {
    "dell-latitude-5420.png": ("Dell Latitude 5420", "#1f2937"),
    "hp-pavilion-15.png": ("HP Pavilion 15", "#3b82f6"),
    "lenovo-thinkpad-e14.png": ("Lenovo ThinkPad E14", "#ef4444"),
    "macbook-air-m1.png": ("MacBook Air M1", "#6b7280"),
    "asus-vivobook-15.png": ("ASUS VivoBook 15", "#10b981"),
    "hp-laserjet-pro-m404n.png": ("HP LaserJet Pro M404n", "#f59e0b"),
    "canon-pixma-ts3320.png": ("Canon Pixma TS3320", "#8b5cf6"),
    "epson-ecotank-l3150.png": ("Epson EcoTank L3150", "#06b6d4"),
    "brother-mfc-l2710dw.png": ("Brother MFC-L2710DW", "#ec4899"),
    "xerox-workcentre-3335.png": ("Xerox WorkCentre 3335", "#14b8a6"),
    "logitech-mx-master-3.png": ("Logitech MX Master 3", "#f97316"),
    "corsair-k95-rgb.png": ("Corsair K95 RGB", "#6366f1"),
    "samsung-ssd-870-evo-1tb.png": ("Samsung SSD 870 EVO 1TB", "#0ea5e9"),
    "wd-my-passport-2tb.png": ("WD My Passport 2TB", "#84cc16"),
    "jbl-quantum-400.png": ("JBL Quantum 400", "#d946ef"),
    "anker-powercore-20000.png": ("Anker PowerCore 20000", "#f43f5e"),
    "iphone-13.png": ("iPhone 13", "#000000"),
    "samsung-galaxy-s22.png": ("Samsung Galaxy S22", "#1f2937"),
    "xiaomi-redmi-note-11.png": ("Xiaomi Redmi Note 11", "#ea580c"),
    "oneplus-10-pro.png": ("OnePlus 10 Pro", "#d4af37"),
    "tecno-camon-19.png": ("Tecno Camon 19", "#7c3aed"),
    "nokia-g50.png": ("Nokia G50", "#0891b2"),
}

def create_image(filename, product_name, color):
    """Create a simple product image with PIL"""
    width, height = 400, 300
    img = Image.new('RGB', (width, height), color=color)
    draw = ImageDraw.Draw(img)
    
    # Draw text
    text = product_name
    text_bbox = draw.textbbox((0, 0), text, font=None)
    text_width = text_bbox[2] - text_bbox[0]
    text_x = (width - text_width) // 2
    text_y = (height - 20) // 2
    draw.text((text_x, text_y), text, fill="white")
    
    # Save to bytes
    img_bytes = io.BytesIO()
    img.save(img_bytes, format='PNG')
    img_bytes.seek(0)
    return img_bytes

def upload_image(filename, img_bytes):
    """Upload image to Supabase Storage (upsert)"""
    url = f"{SUPABASE_URL}/storage/v1/object/{BUCKET_NAME}/{filename}"
    headers = {
        "authorization": f"Bearer {SUPABASE_KEY}",
        "apikey": SUPABASE_KEY,
        "content-type": "image/png",
        "x-upsert": "true",
    }

    response = requests.post(url, data=img_bytes.getvalue(), headers=headers, timeout=30)
    if response.status_code in [200, 201]:
        print(f"✓ Uploaded: {filename}")
        return True

    # Existing file path can return 400 depending on configuration; retry with PUT.
    if response.status_code in [400, 409]:
        put_response = requests.put(url, data=img_bytes.getvalue(), headers=headers, timeout=30)
        if put_response.status_code in [200]:
            print(f"✓ Replaced: {filename}")
            return True
        print(f"✗ Failed to replace {filename}: {put_response.status_code} - {put_response.text}")
        return False

    print(f"✗ Failed to upload {filename}: {response.status_code} - {response.text}")
    return False


def update_product_image_path(product_name, filename):
    """Update products.image_url to products/<filename>."""
    encoded_name = quote(product_name, safe="")
    url = f"{SUPABASE_URL}/rest/v1/products?name=eq.{encoded_name}"
    headers = {
        "authorization": f"Bearer {SUPABASE_KEY}",
        "apikey": SUPABASE_KEY,
        "content-type": "application/json",
        "prefer": "return=representation",
    }
    payload = {"image_url": f"{BUCKET_NAME}/{filename}"}
    response = requests.patch(url, json=payload, headers=headers, timeout=30)
    if response.status_code in [200, 204]:
        print(f"  ↳ DB updated: {product_name}")
        return True
    print(f"  ↳ DB update failed ({product_name}): {response.status_code} - {response.text}")
    return False

def main():
    if not SUPABASE_URL or not SUPABASE_KEY:
        print(
            "Missing credentials. Set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY "
            "(or SUPABASE_KEY) in your environment.",
            file=sys.stderr,
        )
        sys.exit(1)

    print("Starting image upload to Supabase Storage...\n")

    for filename, (product_name, color) in PRODUCTS.items():
        # Create image
        img_bytes = create_image(filename, product_name, color)

        # Upload to Supabase
        if upload_image(filename, img_bytes):
            update_product_image_path(product_name, filename)

    print("\n" + "="*60)
    print("Image upload complete!")
    print("="*60 + "\n")

if __name__ == "__main__":
    main()

File: upload_images.js
#!/usr/bin/env node
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SUPABASE_URL = "https://rrdwvqafkiyhwxvnvhuq.supabase.co";
const SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJyZHd3cWFma2l5aHd4dm52aHVxIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzY2Nzc0OTksImV4cCI6MjA5MjI1MzQ5OX0.clGzqNQToVkikeDgnEBjKVGeZCgq1wYS_JIuodFpCsQ";
const BUCKET_NAME = "products";

const products = {
  "dell-latitude-5420.png": { name: "Dell Latitude 5420", color: "#1f2937" },
  "hp-pavilion-15.png": { name: "HP Pavilion 15", color: "#3b82f6" },
  "lenovo-thinkpad-e14.png": { name: "Lenovo ThinkPad E14", color: "#ef4444" },
  "macbook-air-m1.png": { name: "MacBook Air M1", color: "#6b7280" },
  "asus-vivobook-15.png": { name: "ASUS VivoBook 15", color: "#10b981" },
  "hp-laserjet-pro-m404n.png": { name: "HP LaserJet Pro M404n", color: "#f59e0b" },
  "canon-pixma-ts3320.png": { name: "Canon Pixma TS3320", color: "#8b5cf6" },
  "epson-ecotank-l3150.png": { name: "Epson EcoTank L3150", color: "#06b6d4" },
  "brother-mfc-l2710dw.png": { name: "Brother MFC-L2710DW", color: "#ec4899" },
  "xerox-workcentre-3335.png": { name: "Xerox WorkCentre 3335", color: "#14b8a6" },
  "logitech-mx-master-3.png": { name: "Logitech MX Master 3", color: "#f97316" },
  "corsair-k95-rgb.png": { name: "Corsair K95 RGB", color: "#6366f1" },
  "samsung-ssd-870-evo-1tb.png": { name: "Samsung SSD 870 EVO 1TB", color: "#0ea5e9" },
  "wd-my-passport-2tb.png": { name: "WD My Passport 2TB", color: "#84cc16" },
  "jbl-quantum-400.png": { name: "JBL Quantum 400", color: "#d946ef" },
  "anker-powercore-20000.png": { name: "Anker PowerCore 20000", color: "#f43f5e" },
  "iphone-13.png": { name: "iPhone 13", color: "#000000" },
  "samsung-galaxy-s22.png": { name: "Samsung Galaxy S22", color: "#1f2937" },
  "xiaomi-redmi-note-11.png": { name: "Xiaomi Redmi Note 11", color: "#ea580c" },
  "oneplus-10-pro.png": { name: "OnePlus 10 Pro", color: "#d4af37" },
  "tecno-camon-19.png": { name: "Tecno Camon 19", color: "#7c3aed" },
  "nokia-g50.png": { name: "Nokia G50", color: "#0891b2" },
};

async function generateImage(filename, productData) {
  const svgString = `
    <svg width="400" height="300" xmlns="http://www.w3.org/2000/svg">
      <rect width="400" height="300" fill="${productData.color}"/>
      <text x="200" y="150" font-size="24" font-family="Arial, sans-serif" text-anchor="middle" fill="white" font-weight="bold">
        ${productData.name}
      </text>
    </svg>
  `;

  const outputDir = path.join(__dirname, '.temp-images');
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const outputPath = path.join(outputDir, filename);
  
  await sharp(Buffer.from(svgString))
    .png()
    .toFile(outputPath);
  
  return outputPath;
}

async function uploadImage(filename, filepath) {
  const fileBuffer = fs.readFileSync(filepath);
  
  const url = `${SUPABASE_URL}/storage/v1/object/${BUCKET_NAME}/${filename}`;
  const headers = {
    'authorization': `Bearer ${SUPABASE_KEY}`,
    'content-type': 'image/png',
  };

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers,
      body: fileBuffer,
    });

    if (response.ok || response.status === 200) {
      console.log(`✓ Uploaded: ${filename}`);
      return `${SUPABASE_URL}/storage/v1/object/public/${BUCKET_NAME}/${filename}`;
    } else {
      const text = await response.text();
      console.error(`✗ Failed to upload ${filename}: ${response.status} - ${text}`);
      return null;
    }
  } catch (error) {
    console.error(`✗ Error uploading ${filename}:`, error.message);
    return null;
  }
}

async function main() {
  console.log("Starting image generation and upload...\n");

  const urlMappings = {};

  for (const [filename, productData] of Object.entries(products)) {
    try {
      // Generate image
      const filepath = await generateImage(filename, productData);
      console.log(`Generated: ${filename}`);

      // Upload image
      const publicUrl = await uploadImage(filename, filepath);
      if (publicUrl) {
        urlMappings[productData.name] = publicUrl;
      }

      // Clean up temp file
      fs.unlinkSync(filepath);
    } catch (error) {
      console.error(`Error processing ${filename}:`, error.message);
    }
  }

  // Clean up temp directory
  const tempDir = path.join(__dirname, '.temp-images');
  if (fs.existsSync(tempDir)) {
    fs.rmSync(tempDir, { recursive: true });
  }

  console.log("\n" + "=".repeat(60));
  console.log("Image processing complete!");
  console.log("=".repeat(60) + "\n");

  // Write mappings to JSON for SQL update
  const mappingFile = path.join(__dirname, 'url-mappings.json');
  fs.writeFileSync(mappingFile, JSON.stringify(urlMappings, null, 2));
  console.log(`URL mappings saved to: ${mappingFile}`);
}

main().catch(console.error);
